File: LGPL/CommonSoftware/acsalarm/src/AlarmSystemInterface.cpp
/*******************************************************************************
 * ALMA - Atacama Large Millimeter Array
 * Copyright (c) ESO - European Southern Observatory, 2011
 * (in the framework of the ALMA collaboration).
 * All rights reserved.
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
 *******************************************************************************/
#include "AlarmSystemInterface.h"
#include "logging.h"

using std::vector;
using std::string;
using std::iterator;
using std::auto_ptr;
using acsalarm::FaultState;
using acsalarm::AlarmSystemInterface;

/**
 * Push a collection of fault states.
 * @param states
 *
 * TODO later: @throws ASIException if the fault state collection can not be pushed.
 */
void AlarmSystemInterface::push(vector<FaultState> & states)
{
	ACS_TRACE("AlarmSystemInterface::push(vector<FaultState>)");
	ACE_Guard<ACE_Recursive_Thread_Mutex> guard(m_mutex);
	commonPush(states, false);
}

/**
 * Push a fault state.
 * @param state the fault state change to push.
 *
 * TODO later: @throws ASIException if the fault state can not be pushed.
 */
void AlarmSystemInterface::push(FaultState & state)
{
	ACS_TRACE("AlarmSystemInterface::push(FaultState)");
	ACE_Guard<ACE_Recursive_Thread_Mutex> guard(m_mutex);
	// create a vector and populate with the (single) fault state,
	// to be passed to the buildMessageXML method
	vector<FaultState> states;
	FaultState* st=(FaultState*)&state;
	states.push_back((FaultState)*st);

	commonPush(states, false);
}

/**
 * Push the set of active fault states.
 * @param activeFaults the active fault states.
 *
 * TODO later: @throws ASIException if the fault state active list can not be pushed.
 */
void AlarmSystemInterface::pushActiveList(vector<FaultState> & activeFaults)
{
	ACS_TRACE("AlarmSystemInterface::pushActiveList()");
	ACE_Guard<ACE_Recursive_Thread_Mutex> guard(m_mutex);
	commonPush(activeFaults, true);
}

/**
 * Private method to push a collection of fault states, containing the
 * logic which is common to both the push() and pushActiveList() methods.
 *
 * NOTE: This method is executed only when the CERN alarm system is in use
 *       because ACSAlarmSystemInterfaceProxy redefines the push...()
 *
 * @param states
 * @param backup whether we are sending 'backup' alarms or not. backup alarms
 *        are alarms in the active list that are sent on startup, when the source
 *        starts and periodically according to the expected backup frequency.
 *
 * TODO later: @throws ASIException if the fault state collection can not be pushed.
 */
void AlarmSystemInterface::commonPush(vector<FaultState> & states, bool backup)
{
	ACS_TRACE("AlarmSystemInterface::commonPush()");
	ACE_Guard<ACE_Recursive_Thread_Mutex> guard(m_mutex);
	// create the ASIMessage, supplying the faults which are to be published to the alarm server
	auto_ptr<vector<FaultState> > statesAutoPtr(new vector<FaultState>(states));
	ASIMessage asiMessage(statesAutoPtr);

	// populate the ASIMessage's source timestamp (with the current time)
	auto_ptr<Timestamp> timestampPtr(new Timestamp());
	asiMessage.setSourceTimestamp(timestampPtr);

	// populate the ASIMessage's source name
	asiMessage.setSourceName(sourceName);

	// populate the ASIMessage's source hostname
	asiMessage.setSourceHostname(hostName);

	// set the ASIMessage's backup flag
	asiMessage.setBackup(backup);

	// set the ASIMessage's version
	asiMessage.setVersion(configuration.getASIVersion());

	// publish the ASIMessage to the alarm server
	publishMessage(asiMessage);

	// Log an LM_ALARM message for each fault state in the vector
	for (vector<FaultState>::iterator it = states.begin(); it!=states.end(); ++it/* increment operand is used to move to next element*/) {
	    char msgA[16];
	    sprintf(msgA,"%d",(*it).getCode());
	    string msg="Alarm sent: <";
		msg+=(*it).getFamily();
		msg+=",";
		msg+=(*it).getMember();
		msg+=",";
		msg+=msgA;
		msg+="> ";
		msg+=(*it).getDescriptor();
		ACS_SHORT_LOG((LM_DEBUG, msg.c_str()));
	}
}

File: LGPL/CommonSoftware/acsalarmpy/src/Acsalarmpy/AlarmSystemInterface.py
#! /usr/bin/env python
#*******************************************************************************
# ALMA - Atacama Large Millimiter Array
# (c) National Research Council of Canada, 2008 
# 
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
#
# "@(#) $Id: AlarmSystemInterface.py,v 1.1 2008/10/09 16:11:10 agrimstrup Exp $"
#
# who       when      what
# --------  --------  ----------------------------------------------
# agrimstrup  2008-09-29  created
#

import Acspy.Common.Log as Log
import Acsalarmpy.ASI as ASI
import Acsalarmpy.Timestamp as Timestamp

class AlarmSystemInterface(object):
    """
    Alarm System Interface for Python alarm generators.  It is based
    on CERN's Java implementation.
    """
    def __init__(self, sourceName = None, hostName = None):
        """
        Create an instance of the AlarmSystemInterface.

        Parameters:  sourceName is the name of this source
                     hostName is the name of the computer where the source is running.
        """
        self.logger = Log.getLogger()
        self.sourceName = sourceName
        self.hostName = hostName
        self.configuration = ASI.ASIConfiguration()

    def push(self, state):
        """
        Push fault states to the LASER server.

        Parameters: state is a single fault state or a list of fault states.
        """
        self.logger.logTrace("AlarmSystemInterface: entering push.")
        if isinstance(state, list):
            self._commonPush(state, False)
        else:
            self._commonPush([state], False)
        self.logger.logTrace("AlarmSystemInterface: exiting push.")

    def pushActiveList(self, activeFaults):
        """
        Pushes a list of active fault states to the LASER server.

        Parameters: activeFaults is the list of fault states.
        """
        self.logger.logTrace("AlarmSystemInterface: entering pushActiveList.")
        self._commonPush(activeFaults, True)
        self.logger.logTrace("AlarmSystemInterface: exiting pushActiveList.")

    def close(self):
        """
        Clean up the resources used by this interface.  Must be implemented by
        the subclasses.
        """
        raise Exception("Implement in derived classes")

    def publishMessage(self, msg):
        """
        Send a message to the LASER alarm server.  Must be implemented in the
        subclasses using the desired communication mechanism.

        Parameter:  msg is the ASIMessage to be sent.
        """
        raise Exception("Implement in derived classes")

    def _commonPush(self, states, backup):
        """
        Common method to push a collection of fault states to the LASER server.

        Parameters: states is a list of the fault states to be sent
                    backup is a flag indicating if we are sending backup alarms.
                    Backup alarms are alarms in the active list that are sent on
                    startup, when the source starts and periodically according to
                    the expected backup frequency.
        """
        self.logger.logTrace("AlarmSystemInterface: entering _commonPush.")
        msg = ASI.ASIMessage(states)
        msg.sourceTimestamp = Timestamp.Timestamp()
        msg.sourceName = self.sourceName
        msg.sourceHostname = self.hostName
        msg.backup = backup
        msg.version = self.configuration.asiVersion
        self.publishMessage(msg)
        for s in states:
            self.logger.logAlert('Alarm sent: <%s, %s, %d> %s' % (s.family, s.member, s.code, s.descriptor))
        self.logger.logTrace("AlarmSystemInterface: exiting _commonPush.")
            
#
# ___oOo___
